core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2300,24 +2300,34 @@ object ZIO {
     else {
       val size = as.size
       for {
-        parentId <- ZIO.fiberId
-        causes   <- Ref.make[Cause[E]](Cause.empty)
-        result   <- Promise.make[Unit, Unit]
-        succeed  <- Ref.make(0)
+        parentId       <- ZIO.fiberId
+        causes         <- Ref.make[Cause[E]](Cause.empty)
+        result         <- Promise.make[Nothing, Boolean]
+        failureTrigger <- Promise.make[Unit, Unit]
+        count          <- Ref.make(0)
         fibers <- ZIO.traverse(as) {
                    f(_)
-                     .foldCauseM(c => causes.update(_ && c) *> result.fail(()), _ => {
-                       (succeed.update(_ + 1) >>= { succeed =>
-                         ZIO.when(succeed == size)(result.succeed(()))
+                     .foldCauseM(c => causes.update(_ && c) *> failureTrigger.fail(()), _ => ZIO.unit)
+                     .ensuring {
+                       (count.update(_ + 1) >>= { done =>
+                         ZIO.when(done == size) {
+                           result.complete(failureTrigger.succeed(()))
+                         }
                        }).uninterruptible
```

Comment:
Don't think we need `uninterruptible` here because finalizer in `ensuring` is already uninterruptible.

Reply:
Good catch. I copy/pasted it into the ensuring without engaging my brain.